build/common/bitbake_util.py

```diff
@@ -67,86 +62,84 @@ def __init__(
-        env_vars = self._build_env_variables_string()
-        if env_vars:
-            env_vars = quote(env_vars)
-        command = "{} DISTRO={} MACHINE={} . {} build-{}".format(
-            env_vars,
-            quote(self.distro),
-            quote(self.machine),
-            quote(self.init_env_file),
-            quote(self.distro),
+    def _generate_setup_env_command(self, verbose=False):
+        """Generate a Bash command to set up the BitBake environment."""
+        redirect_str = ""
+        if not verbose:
+            redirect_str = ">/dev/null"
+
+        return ". {} build-{} {}".format(
+            quote(self.init_env_file), quote(self.distro), redirect_str
         )
-        self.run_command(command)
 
-    def run_command(self, command, timeout=None, stdout=True):
+    def run_command(self, command, verbose=False, **kwargs):
         """
         Run a command in the Bitbake environment.
 
+        Runs a command within the BitBake environment using subprocess.run().
+        Most of the subprocess.run options can be used with run_command() and
+        have the same meanings. The exceptions are:
+        * shell: The command is always executed in a Bash shell and must be
+          given as a string rather than a list. The shell option is ignored.
+        * env: The env option works like it does in subprocess.run(), except
+          that the MACHINE and DISTRO environment variables will also be in the
+          environment, along with any that setup-environment sets.
+        * cwd: If cwd is not given then the command is run from the BitBake
+          build directory. If cwd is given and it is a relative path then it
+          will be interpreted relative to the BitBake build directory.
+
         Mandatory args:
         * command (str): the command to run in the environment
 
         Optional args:
-        * timeout (int): how long to wait for the next prompt. None means to
-                         wait indefinitely
-        * stdout (bool): flag to print the output to stdout
+        * verbose (bool): log the command to be executed and its return code to
+                          stdout.
+        * kwargs for subprocess.run().
 
         Return:
-        * output (str): output of the specified command
+        * subprocess.CompletedProcess object.
 
         """
-        print('Running "{}"...'.format(command))
-        output = self._shell.run_command(
-            command, timeout=timeout, async_=False
+        # setup-environment will change the current directory to the BitBake
+        # build dir when it is run, so if the user wants to run a command from
+        # somewhere else then we need to handle that after setup-environment
+        # runs. We can't do that with subprocess.run()'s "cwd" option so add a
+        # "cd" command after sourcing setup-environment.
+        cd_command = ""
+        if "cwd" in kwargs and kwargs["cwd"]:
```

Comment:
This check could possibly be simplified to `if kwargs.get("cwd") is not None:` as `get` returns `None` (or the value of the optional second argument) if the key isn't present.

Reply:
Ah, yes, I'd forgotten about `get`. I think just
```Python
if kwargs.get("cwd"):
```
would be best though. I don't want to treat the empty string (or False, or whatever) as a valid value for cwd.

Reply:
Changed in https://github.com/ARMmbed/mbl-tools/pull/350/commits/5ceee7bed1e502081a336ddded863caf8a61a1fb.